src/utransport.rs

```diff
@@ -58,20 +167,87 @@ pub trait UTransport {
-    ///
-    /// # Returns
-    ///
-    /// An identifier that can be used for [unregistering the listener](Self::unregister_listener) again.
+    /// * `listener` - The listener to invoke. Note that we do not take ownership to communicate
+    ///                that a caller should keep a copy to be able to call `unregister_listener()`
     ///
     /// # Errors
     ///
     /// Returns an error if the listener could not be registered.
+    ///
+    /// # Examples
+    ///
+    /// ## Registering a listener
+    ///
+    /// ```
+    /// use std::sync::Arc;
+    /// use up_rust::UListener;
+    /// # use up_rust::{UMessage, UStatus, UTransport, UUri};
+    /// # use async_trait::async_trait;
+    /// #
+    /// # pub struct MyTransport;
+    /// #
+    /// # impl MyTransport {
+    /// #     pub fn new()->Self {
+    /// #         Self
+    /// #     }
+    /// # }
+    /// #
+    /// # #[async_trait]
+    /// # impl UTransport for MyTransport {
+    /// #     async fn send(&self, _message: UMessage) -> Result<(), UStatus> {
+    /// #         todo!()
+    /// #     }
+    /// #
+    /// #     async fn receive(&self, _topic: UUri) -> Result<UMessage, UStatus> {
+    /// #         todo!()
+    /// #     }
+    /// #
+    /// #     async fn register_listener(&self, _topic: UUri, _listener: &Arc<dyn UListener>) -> Result<(), UStatus> {
+    /// #         Ok(())
+    /// #     }
+    /// #
+    /// #     async fn unregister_listener(&self, _topic: UUri, _listener: Arc<dyn UListener>) -> Result<(), UStatus> {
+    /// #         Ok(())
+    /// #     }
+    /// # }
+    /// #
+    /// # #[derive(Clone)]
+    /// # pub struct MyListener;
+    /// #
+    /// # impl MyListener {
+    /// #     pub fn new() -> Self {
+    /// #         Self
+    /// #     }
+    /// # }
+    /// #
+    /// # #[async_trait]
+    /// # impl UListener for MyListener {
+    /// #     async fn on_receive(&self, msg: UMessage) {
+    /// #         todo!()
+    /// #     }
+    /// #
+    /// #     async fn on_error(&self, err: UStatus) {
+    /// #         todo!()
+    /// #     }
+    /// # }
+    /// #
+    /// # let my_transport = MyTransport::new();
+    /// # let my_uuri = UUri::default();
+    ///
+    /// // hang onto this listener...
+    /// let my_listener: Arc<dyn UListener> = Arc::new(MyListener::new());
+    /// // ...send a clone through when registering...
+    /// let my_listener_for_transport = my_listener.clone();
+    /// let register_result = my_transport.register_listener(my_uuri.clone(), &my_listener_for_transport);
+    ///
+    /// // ...and use the original we hung onto when unregistering
+    /// let unregister_result = my_transport.unregister_listener(my_uuri, my_listener);
+    /// ```
     async fn register_listener(
         &self,
         topic: UUri,
-        listener: Box<dyn Fn(Result<UMessage, UStatus>) + Send + Sync + 'static>,
-    ) -> Result<String, UStatus>;
+        listener: &Arc<dyn UListener>,
```

Comment:
I do not understand why you have turned this into a *reference to a reference*. Can you elaborate?